File: python/tools/domains/groups.py
# Copyright since 2025 Mifos Initiative
# This Source Code Form is subject to the terms of the Mozilla Public
# License, v. 2.0. If a copy of the MPL was not distributed with this
# file, You can obtain one at http://mozilla.org/MPL/2.0/.

import datetime

from langchain_core.tools import tool

from tools.mcp_adapter import fineract_client

# --- GROUP OPERATIONS ---

@tool
def list_groups(office_id: int = None):
    """Answers: 'Show me all lending groups' or 'List groups in office #1'"""
    endpoint = "groups"
    if office_id:
        endpoint += f"?officeId={office_id}"
    print("[Tool] Fetching Groups...")
    return fineract_client.execute_get(endpoint)

@tool
def get_group(group_id: int):
    """Answers: 'Show me details for group #12'"""
    print(f"[Tool] Fetching details for Group #{group_id}...")
    return fineract_client.execute_get(f"groups/{group_id}?associations=all")

@tool
def create_group(name: str, office_id: int, external_id: str = None):
    """Answers: 'Create a new lending group named "Alpha Group" in office 1'"""
    print(f"[Tool] Creating Group '{name}'...")
    payload = {
        "name": name,
        "officeId": office_id,
        "active": False,
        "externalId": external_id
    }
    return fineract_client.execute_post("groups", payload)

@tool
def activate_group(group_id: int):
    """Answers: 'Activate this pending group'"""
    print(f"[Tool] Activating Group #{group_id}...")
    today = datetime.datetime.now().strftime("%d %B %Y")
    payload = {
        "activationDate": today,
        "dateFormat": "dd MMMM yyyy",
        "locale": "en"
    }
    return fineract_client.execute_post(f"groups/{group_id}?command=activate", payload)

@tool
def add_group_member(group_id: int, client_id: int):
    """Answers: 'Add client #5 to group #12'"""
    print(f"[Tool] Adding Client #{client_id} to Group #{group_id}...")
    payload = {"clientMembers": [client_id]}
    return fineract_client.execute_post(f"groups/{group_id}?command=associateClients", payload)

# --- CENTER OPERATIONS ---

@tool
def list_centers(office_id: int = None):
    """Answers: 'List all centers'"""
    endpoint = "centers"
    if office_id:
        endpoint += f"?officeId={office_id}"
    print("[Tool] Fetching Centers...")
    return fineract_client.execute_get(endpoint)

@tool
def get_center(center_id: int):
    """Answers: 'Show me details for center #5'"""
    print(f"[Tool] Fetching details for Center #{center_id}...")
    return fineract_client.execute_get(f"centers/{center_id}?associations=groupMembers,collectionSheet")

@tool
def create_center(name: str, office_id: int, external_id: str = None):
    """Answers: 'Create a new center named "Main Center" in office 1'"""
    print(f"[Tool] Creating Center '{name}'...")
    payload = {
        "name": name,
        "officeId": office_id,
        "active": False,
        "externalId": external_id
    }
    return fineract_client.execute_post("centers", payload)

File: rust/src/domains/groups.rs
// Copyright since 2025 Mifos Initiative
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at http://mozilla.org/MPL/2.0/.

use crate::adapter::FineractAdapter;
use rmcp::schemars;
use rmcp::{ErrorData as McpError, model::{CallToolResult, Content}};
use serde_json::json;
use chrono::Local;

pub fn to_result(val: serde_json::Value) -> Result<CallToolResult, McpError> { Ok(CallToolResult::success(vec![Content::text(serde_json::to_string_pretty(&val).unwrap_or_default())])) }
pub fn to_err(err: anyhow::Error) -> McpError { McpError::internal_error(err.to_string(), None) }
pub fn today() -> String { Local::now().format("%d %B %Y").to_string() }

#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct ListGroupsReq { pub office_id: Option<i64> }
#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct GroupIdReq { pub group_id: i64 }
#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct CreateGroupReq { pub name: String, pub office_id: i64, pub external_id: Option<String>, pub client_members: Option<Vec<i64>> }
#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct AddGroupMemberReq { pub group_id: i64, pub client_id: i64 }
#[derive(Debug, serde::Deserialize, schemars::JsonSchema)]
pub struct CenterIdReq { pub center_id: i64 }

pub async fn list_groups(adapter: &FineractAdapter, req: ListGroupsReq) -> Result<CallToolResult, McpError> {
    let mut endpoint = "groups".to_string();
    if let Some(id) = req.office_id { endpoint.push_str(&format!("?officeId={}", id)); }
    let res = adapter.execute_get(&endpoint, None).await.map_err(to_err)?;
    to_result(res)
}

pub async fn get_group(adapter: &FineractAdapter, req: GroupIdReq) -> Result<CallToolResult, McpError> {
    let res = adapter.execute_get(&format!("groups/{}?associations=all", req.group_id), None).await.map_err(to_err)?;
    to_result(res)
}

pub async fn create_group(adapter: &FineractAdapter, req: CreateGroupReq) -> Result<CallToolResult, McpError> {
    let mut payload = json!({ "name": req.name, "officeId": req.office_id, "active": false, "externalId": req.external_id });
    if let Some(members) = req.client_members { payload["clientMembers"] = json!(members); }
    let res = adapter.execute_post("groups", &payload).await.map_err(to_err)?;
    to_result(res)
}

pub async fn activate_group(adapter: &FineractAdapter, req: GroupIdReq) -> Result<CallToolResult, McpError> {
    let payload = json!({ "activationDate": today(), "dateFormat": "dd MMMM yyyy", "locale": "en" });
    let res = adapter.execute_post(&format!("groups/{}?command=activate", req.group_id), &payload).await.map_err(to_err)?;
    to_result(res)
}

pub async fn add_group_member(adapter: &FineractAdapter, req: AddGroupMemberReq) -> Result<CallToolResult, McpError> {
    let payload = json!({ "clientMembers": [req.client_id] });
    let res = adapter.execute_post(&format!("groups/{}?command=associateClients", req.group_id), &payload).await.map_err(to_err)?;
    to_result(res)
}

pub async fn list_centers(adapter: &FineractAdapter, req: ListGroupsReq) -> Result<CallToolResult, McpError> {
    let mut endpoint = "centers".to_string();
    if let Some(id) = req.office_id { endpoint.push_str(&format!("?officeId={}", id)); }
    let res = adapter.execute_get(&endpoint, None).await.map_err(to_err)?;
    to_result(res)
}

pub async fn get_center(adapter: &FineractAdapter, req: CenterIdReq) -> Result<CallToolResult, McpError> {
    let res = adapter.execute_get(&format!("centers/{}?associations=groupMembers,collectionSheet", req.center_id), None).await.map_err(to_err)?;
    to_result(res)
}

pub async fn create_center(adapter: &FineractAdapter, req: CreateGroupReq) -> Result<CallToolResult, McpError> {
    let payload = json!({ "name": req.name, "officeId": req.office_id, "active": false, "externalId": req.external_id });
    let res = adapter.execute_post("centers", &payload).await.map_err(to_err)?;
    to_result(res)
}
